phabricatoremails/source.py

```diff
@@ -0,0 +1,80 @@
+import json
+import pathlib
+from dataclasses import dataclass
+from typing import Optional
+
+import requests
+from requests import RequestException
+
+
+class PhabricatorException(Exception):
+    """An error that occurred while communicating with Phabricator"""
+
+    pass
+
+
+@dataclass
+class PhabricatorSource:
+    """Fetches feed information directly from the Phabricator API"""
+
+    server: str
+    token: str
+    override_limit: Optional[int]
+
+    def _request(self, endpoint, **query_parameters):
+        query_parameters = {k: v for k, v in query_parameters.items() if v}
+        url = "{}/api/{}".format(self.server, endpoint)
+
+        body = {
+            "params": json.dumps(
+                {"__conduit__": {"token": self.token}, **query_parameters}
+            )
+        }
+
+        try:
+            result = requests.post(url, data=body)
+        except RequestException:
+            raise PhabricatorException("Could not communicate with Phabricator")
+
+        if result.status_code != 200:
+            raise PhabricatorException(
+                "Phabricator request had incorrect status of {}. "
+                "The full response is: {}".format(result.status_code, result.text,)
+            )
+
+        body = result.json()
+
+        if body["error_code"]:
+            raise PhabricatorException(
+                "Phabricator request had an error: {}".format(body)
+            )
+
+        return body
+
+    def fetch_feed_end(self):
+        body = self._request("feed.for_email.status")
```

Comment:
Can't remember if I commented on this earlier, but I think these conduit API strings should go somewhere where they are easy to see as a whole - e.g. an enum. It will make it easier to first see all the possible API calls that we can make, and second, if the API changes it will be a good place to look for the definitions used in this service.

Reply:
> It will make it easier to first see all the possible API calls that we can make

Since `PhabricatorSource` encapsulates all communication with Phabricator, and has a pattern of exposing each API call as a function, this information is still similarly find-able.

> if the API changes it will be a good place to look for the definitions used in this service.

If the API changes, the name of the method is just one part of the API: there's the structure of the request and response body, headers, and a host of other details as well. Shouldn't we keep the details for each API call together?

Reply:
> If the API changes, the name of the method is just one part of the API: there's the structure of the request and response body, headers, and a host of other details as well. Shouldn't we keep the details for each API call together?

I think that the location / URL is universal among all endpoints, and in your implementation it's the only thing using strings. Fair point, though as a matter of preference I think it's cleaner to not have arbitrary strings used in your calls (e.g. makes it harder to test). Remember that these values are going to probably be repeated in your tests (which is when this need for a mapping will be more essential).